subsys/bluetooth/scan.c

```diff
@@ -1346,3 +1460,52 @@ int bt_scan_params_set(struct bt_le_scan_param *scan_param)
 
 	return 0;
 }
+
+int bt_scan_attempts_filter_device_generic_add(const bt_addr_le_t *addr,
+					       bool force)
+{
+	int err = 0;
+	struct conn_attempts_filter *filter = &bt_scan.attempts_filter;
+
+	k_mutex_lock(&conn_attempts_mutex, K_FOREVER);
+
+	/* Check if device is already in the filter array. */
+	for (size_t i = 0; i < filter->count; i++) {
+		struct conn_attempts_device *device = &filter->device[i];
+
+		if (bt_addr_le_cmp(addr, &device->addr) == 0) {
+			LOG_DBG("Device is already in the filter arrary");
+			goto out;
+		}
+	}
+
+	if (filter->count >= ARRAY_SIZE(filter->device)) {
+		if (force) {
+			if (!filter->oldest_device) {
+				filter->oldest_device = filter->device;
+			}
+
+			attempts_filter_force_add(filter, addr);
+		} else {
+			LOG_ERR("No memory for the new device filter");
+			err = -ENOMEM;
+		}
+
+		goto out;
+	}
+
+	memcpy(&filter->device[filter->count].addr, addr, sizeof(*addr));
+	filter->count++;
+
+out:
+	k_mutex_unlock(&conn_attempts_mutex);
+
+	return err;
+}
+
+void bt_scan_attempts_filter_clean(void)
+{
+	k_mutex_lock(&conn_attempts_mutex, K_FOREVER);
+	memset(&bt_scan.attempts_filter, 0, sizeof(bt_scan.attempts_filter));
```

Comment:
Maybe we could set the `oldest_device` pointer here and during the module initialization. Then we could always use it instead of array[count] when adding new device.

Reply:
Done. I still need to use the count parameter :)

Reply:
Instead of a pointer to the oldest device, I think it is more logical to keep an index to it. that way you don't have to set a valid pointer in the beginning.